src/coreclr/jit/lclvars.cpp

```diff
@@ -275,7 +275,26 @@ void Compiler::lvaInitTypeRef()
 
         if ((corInfoTypeWithMod & CORINFO_TYPE_MOD_PINNED) != 0)
         {
-            if ((corInfoType == CORINFO_TYPE_CLASS) || (corInfoType == CORINFO_TYPE_BYREF))
+            CORINFO_CLASS_HANDLE refClsHnd = NO_CLASS_HANDLE;
+            // don't run in crossgen to avoid invalid type info
+            if (corInfoType == CORINFO_TYPE_BYREF && !opts.IsReadyToRun())
+            {
+                CORINFO_CLASS_HANDLE clsHnd = info.compCompHnd->getArgClass(&info.compMethodInfo->locals, localsSig);
+
+                CORINFO_CLASS_HANDLE realClsHnd = NO_CLASS_HANDLE;
+                if (info.compCompHnd->getChildType(clsHnd, &realClsHnd) == CORINFO_TYPE_VALUECLASS)
+                {
+                    refClsHnd = realClsHnd;
+                }
+            }
+
+            // ignore pinned on ref structs
+            if ((refClsHnd != NO_CLASS_HANDLE) &&
+                ((info.compCompHnd->getClassAttribs(refClsHnd) & CORINFO_FLG_BYREF_LIKE) != 0))
+            {
+                JITDUMP("Ignoring pinned on local V%02u for byref-like type %s\n", varNum, eeGetClassName(refClsHnd));
```

Comment:
@jkotas We were having a discussion on Discord whether enforcing `ref struct` `ref`s not pointing to the heap is an acceptable assumption for the JIT to make, could you give your opinion on the safety of such assumption?

Reply:
The concern is about aliasing. Before this change the JIT's aliasing model allows roundtripping any `ref A` as a `ref B`. The optimization in this PR is not valid unless we consider that to be UB when A and B do not match in byref likeness.